### PR TITLE
#2 Station search via device positioning services

### PR DESCRIPTION
**Pull request by jheubuch**
This pull request implements the location-based search of the nearest station -> Issue #2

Therefore an extra button was added in the station input field:
image

When clicking the button the following will happen:
- if the user clicks the button the first time, he will be asked to grant the permission to use the positioning services
- if the permission was granted and the positioning was successful, the most closest station will be queried and selected for the selections of the trains, buses etc
- if the permission was denied, an error message will be displayed

### CODE REVIEW

**Comment by MrKrisKrisu** [comment]
This is more a question for the Traewelling team: @HerrLevin @jeyemwey 

The variables and function names are partly written in correct CamelCase throughout the whole code (not just this PR), but sometimes with large letters at the beginning. I even found underlined names. It may make sense to standardize this, it doesn't look nice.

File: app/Http/Controllers/FrontendTransportController.php [hunk 1]
```diff
@@ -42,6 +42,22 @@ public function TrainStationboard(Request $request) {
         );
     }
 
+    public function StationByCoordinates(Request $request)
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/FrontendTransportController.php:45]: I recommend renaming it "getStationByCoordinates", even if other method names are in this format. See my general answer to your pull request for further information.

File: app/Http/Controllers/FrontendTransportController.php [hunk 2]
```diff
@@ -42,6 +42,22 @@ public function TrainStationboard(Request $request) {
...
+    {
+        $NearestStation = TransportBackend::StationByCoordinates($request->latitude, $request->longitude);
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/FrontendTransportController.php]: I recommend renaming the variable to "$nearestStation"

File: app/Http/Controllers/TransportController.php [hunk 3]
```diff
@@ -96,6 +96,15 @@ public static function FastTripAccess($departure, $lineName, $number, $when)
         return null;
     }
 
+    public static function StationByCoordinates($latitude, $longitude)
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/TransportController.php:99]: I recommend renaming it "getStationByCoordinates"

File: app/Http/Controllers/TransportController.php [hunk 4]
```diff
@@ -96,6 +96,15 @@ public static function FastTripAccess($departure, $lineName, $number, $when)
...
+    {
+        $client = new Client(['base_uri' => config('trwl.db_rest')]);
+        $response = $client->request('GET', "stops/nearby?latitude=$latitude&longitude=$longitude&results=1");
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/TransportController.php:102]: Maybe it would be beautiful to use "http_build_query()" function, but it's not neccessary.

File: resources/lang/de/stationboard.php [hunk 5]
```diff
@@ -23,5 +23,7 @@
     "submit-search" => "Suchen",
     "where-are-you" => "Wo bist Du?",
     "last-stations" => "Letzte Stationen",
-    "next-stop" => "Nächster Halt"
+    "next-stop" => "Nächster Halt",
+    "search-by-location" => "Standortbasierte Suche",
+    "position-unavailable" => "Postionsbestimmung nicht möglich!"
```
Comment (MrKrisKrisu) [inline resources/lang/de/stationboard.php]: There's a spelling mistake at "Postionsbestimmung"

File: resources/views/layouts/app.blade.php [hunk 6]
```diff
@@ -169,7 +169,7 @@ class="fas fa-tools"></i> {{__('menu.admin')}}</a>
                     <span class="footer-nav-link">/ <a href="{{ route('blog.all') }}">{{ __('menu.blog') }}</a></span>
                 </p>
                 <p class="mb-0">{!! __('menu.developed') !!}</p>
-                <p class="mb-0">&copy; 2019 Tr&auml;welling</p>
+                <p class="mb-0">&copy; 2020 Tr&auml;welling</p>
```
Comment (MrKrisKrisu) [inline resources/views/layouts/app.blade.php:172]: Maybe it would be better to change the year to {{date('Y')}} to prevent these changes.
Even if this is not the primary part of the PR.

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/FrontendTransportController.php (hunk 2 above)
Comment (jeyemwey) [inline app/Http/Controllers/FrontendTransportController.php]: Can we add [input validation](https://laravel.com/docs/5.8/validation) for this?

File: routes/web.php [hunk 7]
```diff
@@ -256,6 +256,11 @@
         'as'    => 'trains.stationboard',
     ]);
 
+    Route::get('/trains/nearby', [
+        'uses'  => 'FrontendTransportController@StationByCoordinates',
+        'as'    => 'trains.nearby',
+    ]);
+
```
Comment (jeyemwey) [inline routes/web.php:263]: Can we add an API request for this? Then we need to adjust the `swagger-api.yml` too. Maybe @HerrLevin can help? This is non-blocking from my side, and can be done later.

File: app/Http/Controllers/FrontendTransportController.php [hunk 8]
```diff
@@ -42,6 +42,22 @@ public function TrainStationboard(Request $request) {
...
+            return redirect()->back()->with('error', __('controller.transport.no-station-found'));
+        }
+
+        return view('stationboard', [
+            'station' => $NearestStation['station'],
+            'departures' => $NearestStation['departures'],
+            'when' => $NearestStation['when'],
+            'request' => $request,
+            'latest' => \App\Http\Controllers\TransportController::getLatestArrivals(Auth::user())
+        ]);
```
Comment (jeyemwey) [inline app/Http/Controllers/FrontendTransportController.php]: I am not sure if this is the best way here.

I'd suggest to redirect to `https://traewelling.de/trains/stationboard?provider=train&station=<stationname>` with `302 Found`? If we later change the stationboard view, we would have to modify its input params in two places which can cause bugs.

**Review by jeyemwey — DISMISSED** [review]
Hi @derheubi, thank you _soo_ much for your contribution to Träwelling. I have posted a few remarks, but I'm sure those can be resolved quickly.

Can you (or someone from our core team) please add unit tests for the new Endpoint? I'm especially interested in edge cases like Null Island or the Poles, and far away places like New York, or Beijing. However, it is also possible to add valid tests like Berlin Hbf, Moskau Central or Frankfurt Flughafen Fernbf.

File: app/Http/Controllers/FrontendTransportController.php (hunk 8 above)
Comment (jheubuch) [reply app/Http/Controllers/FrontendTransportController.php]: Basically, I agree. But is there any way in PHP that there is no need to hard-code "https://traewelling....." or wouldn't it just be enough to "outsource" the view()-call into another function?

Comment (jeyemwey) [reply app/Http/Controllers/FrontendTransportController.php]: You would do something like...

```php
return redirect(route("stationboard", [
  'provider' => 'train',
  'station' => $stationname
]));
```

... which is a Laravel function.

File: app/Http/Controllers/FrontendTransportController.php (hunk 2 above)
Comment (jeyemwey) [reply app/Http/Controllers/FrontendTransportController.php]: You can validate the input like this:

```suggestion
    {
	    $request->validate([
	        'latitude' => 'required|numeric|min:-180|max:180',
	        'longitude' => 'required|numeric|min:-90|max:90'
	    ]);
        $NearestStation = TransportBackend::StationByCoordinates($request->latitude, $request->longitude);
```

**Comment by jheubuch** [comment]
> Can you (or someone from our core team) please add unit tests for the new Endpoint? I'm especially interested in edge cases like Null Island or the Poles, and far away places like New York, or Beijing. However, it is also possible to add valid tests like Berlin Hbf, Moskau Central or Frankfurt Flughafen Fernbf.

The question is, if it is of need to have unit tests for this. It's a core function of the db-rest API. Otherwise the db-rest API needs to be mocked and (I'm sorry :D) I have absolutely no idea how to do this in PHP.

File: app/Http/Controllers/FrontendTransportController.php [hunk 9]
```diff
@@ -44,18 +44,20 @@ public function TrainStationboard(Request $request) {
 
     public function StationByCoordinates(Request $request)
     {
+        $request->validate([
```
Comment (MrKrisKrisu) [inline app/Http/Controllers/FrontendTransportController.php]: You have to save the result from ->validate in an variable to use the validated input.

for example: 
47: $validated = $request->validate([......
...
52: ....ates($validated['latitude'], $validated['longitude']);

**Review by MrKrisKrisu — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/FrontendTransportController.php (hunk 9 above)
Comment (jeyemwey) [reply app/Http/Controllers/FrontendTransportController.php]: Sorry, my bad.

**Review by MrKrisKrisu — APPROVED** [review]

**Comment by jeyemwey** [comment]
> The question is, if it is of need to have unit tests for this. It's a core function of the db-rest API. Otherwise the db-rest API needs to be mocked and (I'm sorry :D) I have absolutely no idea how to do this in PHP.

We do not use mocking for any of the API calls, if db-rest breaks, our CI is going down too. I can add the tests, if you want.

**Comment by jheubuch** [comment]
> > The question is, if it is of need to have unit tests for this. It's a core function of the db-rest API. Otherwise the db-rest API needs to be mocked and (I'm sorry :D) I have absolutely no idea how to do this in PHP.
> 
> We do not use mocking for any of the API calls, if db-rest breaks, our CI is going down too. I can add the tests, if you want.

That would be very nice! Maybe I can do the future upcoming unit tests on my own then 😅

**Comment by HerrLevin** [comment]
Since we're planning to move the frontend to a JS-based API-client in the future, could you maybe implement this as an API-endpoint?
If not, that's no problem. It's just easier for us. :sweat_smile:

**Comment by jheubuch** [comment]
> Since we're planning to move the frontend to a JS-based API-client in the future, could you maybe implement this as an API-endpoint?
> If not, that's no problem. It's just easier for us. 😅

For sure! I was looking at this topic anyway (it was mentioned by @jeyemwey in a review above)

**Comment by jheubuch** [comment]
@HerrLevin I tried adding an API endpoint but actually I had no chance to test it (I get an error when sending an API request):
image

I also added the endpoint to the api-swagger-v0.yml (swagger.io says, the configuration is valid)

**Comment by jeyemwey** [comment]
I have written some tests now for the frontend. @derheubi, can you check the "Allow edits from maintainers" checkbox? https://tighten.co/blog/adding-commits-to-a-pull-request/

**Comment by jheubuch** [comment]
> I have written some tests now for the frontend. @derheubi, can you check the "Allow edits from maintainers" checkbox? https://tighten.co/blog/adding-commits-to-a-pull-request/

Done

**Review by jeyemwey — APPROVED** [review]
Looks good to me now.

**Comment by MrKrisKrisu** [comment]
Before merging: Can we speak about my comments about unified conventions?
@derheubi has used CamelCase Names which started with an uppercase letter. In the whole traewelling code there is no unified procedure for this. Sometimes there is a "getObject" or "ObjectName" Identifier.

It's non blocking for me here, but it needs to be discussed.

//Edit: Okay, 20 seconds too late. Merged. xD

**Comment by jeyemwey** [comment]
Sorry 😂 I have resolved tha sometimes, so there are only few issues left.
